src/vs/platform/contextkey/common/contextkey.ts

```diff
@@ -241,6 +269,370 @@ export abstract class ContextKeyExpr {
+							case 'true':
+								return ContextKeyExpr.not(key);
+							case 'false':
+								return ContextKeyExpr.has(key);
+							default:
+								return ContextKeyExpr.notEquals(key, right);
+						}
+					}
+
+					// TODO: ContextKeyExpr.smaller(key, right) accepts only `number` as `right` AND during eval of this node, we just eval to `false` if `right` is not a number
+					// consequently, package.json linter should _warn_ the user if they're passing undesired things to ops
+					case TokenType.Lt: return ContextKeySmallerExpr.create(key, right);
+					case TokenType.LtEq: return ContextKeySmallerEqualsExpr.create(key, right);
+					case TokenType.Gt: return ContextKeyGreaterExpr.create(key, right);
+					case TokenType.GtEq: return ContextKeyGreaterEqualsExpr.create(key, right);
+
+					case TokenType.In: return ContextKeyExpr.in(key, right);
+
+					default:
+						throw illegalState(`must've matched the op with this._match() call above`);
+				}
+			}
+
+			return ContextKeyExpr.has(key);
+		} else {
+			throw this._errExpectedButGot(`'true', 'false', '(', KEY, KEY '=~' regex, KEY [ ('==' | '!=' | '<' | '<=' | '>' | '>=' | 'in' | 'not' 'in') value ]`, this._peek());
+		}
+	}
+
+	private _value(): string { // TODO@ulugbekna: match all at once and then `switch` on the type
+		if (this._match(TokenType.Str, TokenType.QuotedStr)) {
+			return this._previous().lexeme!;
+		} if (this._match(TokenType.True)) {
+			return 'true';
+		} if (this._match(TokenType.False)) {
+			return 'false';
+		} if (this._match(TokenType.In)) { // we support `in` as a value, e.g., "when": "languageId == in" - exists in existing extensions
+			return 'in';
+		} else {
+			return ''; // this allows "when": "foo == " which's used by existing extensions
+		}
+	}
+
+	// careful: this can throw if current token is the initial one (ie index = 0)
+	private _previous() {
+		return this._tokens[this._current - 1];
+	}
+
+	private _match(...tokens: TokenType[]) {
+		for (const token of tokens) {
+			if (this._check(token)) {
+				this._advance();
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	private _advance() {
+		if (!this._isAtEnd()) {
+			this._current++;
+		}
+		return this._previous();
+	}
+
+	private _consume(type: TokenType, message: string) {
+		if (this._check(type)) {
+			return this._advance();
+		}
+
+		throw this._errExpectedButGot(message, this._peek());
+	}
+
+	private _errExpectedButGot(expected: string, got: Token) {
+		return this._error(`Expected ${expected} but got '${this._nameToken(got)}' at offset ${got.offset}.`);
+	}
+
+	private _errUnexpected(token: Token) {
+		return this._error(`Unexpected '${this._nameToken(token)}' at offset ${token.offset}.`);
+	}
+
+	// TODO@ulugbekna: the whole error reporting needs reworking - especially, when we introduce it to package.json linting
+	private _error(errMsg: string) {
+		this._parsingErrors.push(errMsg);
+		return new ParseError();
+	}
+
+	private _nameToken(token: Token): string {
```

Comment:
This could be extracted to a function in the same file where `Token` is defined (because it doesn't seem to use any properties).